Add code to parse 8:367:24 message.

andyvan-trabus opened the pull request:
This is the smallest message.

schwehr commented on src/libais/ais.h:
```diff
@@ -2433,6 +2433,23 @@ class Ais27 : public AisMsg {
+// SSW Satellite Ship Weather 1-Slot Version
+class Ais8_367_24 : public Ais8 {
+ public:
+  int version;
```
I'm switching to default member initializer style, so please change to `int version = 0;` and similar for all the data members.

schwehr commented on src/libais/ais8.cpp:
```diff
@@ -629,5 +629,34 @@ Ais8_1_31::Ais8_1_31(const char *nmea_payload, const size_t pad)
+// SSW FI24 Satellite Ship Weather Small - Less than 1-Slot Version
+Ais8_367_24::Ais8_367_24(const char *nmea_payload, const size_t pad)
+    : Ais8(nmea_payload, pad), version(0),
+      utc_hour(), utc_min(), position(), pressure(0.0) {
```
Drop the version, utc, pos and pressure initializers here

andyvan-trabus replied:
Done

schwehr commented on src/libais/ais8.cpp:
```diff
@@ -629,5 +629,34 @@ Ais8_1_31::Ais8_1_31(const char *nmea_payload, const size_t pad)
+  position = bits.ToAisPoint(70, 49);
+  pressure = bits.ToUnsignedInt(119, 9) + 799;  // hPa
+
+  // NOTE: If message comes in at 144, then there will be 16 bits remaining.
```
What is in those 16 bits?  It is required to be zero, is it noise?

andyvan-trabus replied:
I've copied you on email to Greg Johnson asking about this.  Most messages come in with 128 bits as expected, but sometimes otherwise valid looking messages come in with an extra 16 bits, which always seem to be zero.

I'm guessing this is an encoder bug.

andyvan-trabus replied:
I've moved this code to asi8_367.cpp.

schwehr commented on src/libais/ais.h:
```diff
@@ -2433,6 +2433,23 @@ class Ais27 : public AisMsg {
 };
 ostream& operator<< (ostream &o, const Ais27 &msg);
 
+//
```
Please comment here where to find the message specification

schwehr reviewed (changes requested):
Please add a C++ test to https://github.com/schwehr/libais/blob/master/src/test/ais8_367_test.cc

andyvan-trabus commented on src/libais/ais8_367.cpp:
```diff
@@ -163,4 +165,51 @@ Ais8_367_22::Ais8_367_22(const char *nmea_payload, const size_t pad)
   status = AIS_OK;
 }
 
+Ais8_367_22::~Ais8_367_22() {
```
I don't think I added this code, it looks like it came in from somebody else's changes that were put into this branch.

andyvan-trabus replied:
I've removed this code.

andyvan-trabus commented:
I'm sorry, but I'm unclear as to the next step.  I believe I've addressed all of the review comments, and I've merged in changes that other people have made to this branch (why?).

schwehr commented on src/libais/ais8_367.cpp:
```diff
@@ -163,4 +165,44 @@ Ais8_367_22::Ais8_367_22(const char *nmea_payload, const size_t pad)
+    return;
+  }
+
+  if ((num_bits != 128) && (num_bits != 144)) {
```
I'm not going to accept non-compliant message handling.  Please remove the 144 bit code paths.

andyvan-trabus replied:
Ok, that just means I'll have to use an unofficial copy of this library for my use.  The encoder was written by Interlake, and apparently the developer has retired.

schwehr commented:
> I'm sorry, but I'm unclear as to the next step. I believe I've addressed all of the review comments, and I've merged in changes that other people have made to this branch (why?).

Apologies!  I missed hitting finish on the review button.  I do 99.9% of my code reviews outside of github.

The key thing that needs to happen is removing the code related to the non-compliant messages.  Devices sending bad messages should be fix or retired.  If someone is looking for deep bit inspection, then using things like noaadata + bitvector is more appropriate.

schwehr commented:
Build failure:

```
build/temp.linux-x86_64-3.8/src/libais/ais8_366.o -std=c++11
    g++-6 -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/opt/hostedtoolcache/Python/3.8.11/x64/include/python3.8 -c src/libais/ais8_367.cpp -o build/temp.linux-x86_64-3.8/src/libais/ais8_367.o -std=c++11
    src/libais/ais8_367.cpp:168:27: error: definition of implicitly-declared ‘virtual libais::Ais8_367_22::~Ais8_367_22()’
     Ais8_367_22::~Ais8_367_22() {
                               ^
    error: command 'g++-6' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /opt/hostedtoolcache/Python/3.8.11/x64/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-volbnmpd/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-volbnmpd/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-lxm1sy9r/install-record.txt --single-version-externally-managed --compile --install-headers /opt/hostedtoolcache/Python/3.8.11/x64/include/python3.8/libais Check the logs for full command output.
```

andyvan-trabus commented:
Okay, I think I've made all of the requested changes.  And, I wasn't being snarky, I'm not sure of the workflow here.  Once I push a commit with the changes, is there anything else I need to do to keep this moving?

schwehr commented:
> Okay, I think I've made all of the requested changes. And, I wasn't being snarky, I'm not sure of the workflow here. Once I push a commit with the changes, is there anything else I need to do to keep this moving?

I didn't see any snark.  I'm not totally sure of the workflow either... sigh.  The change looks good to me.  Thanks for sticking with it!
